go/signedexchange/signedexchange.go

```diff
@@ -0,0 +1,187 @@
+package signedexchange
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/WICG/webpackage/go/signedexchange/cbor"
+	"github.com/WICG/webpackage/go/signedexchange/mice"
+)
+
+type Exchange struct {
+	// Request
+	requestUri *url.URL
+
+	// Response
+	responseStatus int
+	responseHeader http.Header
+
+	// Payload
+	payload []byte
+}
+
+func NewExchange(uri *url.URL, status int, headers http.Header, payload []byte, miRecordSize int) (*Exchange, error) {
+	e := &Exchange{
+		requestUri:     uri,
+		responseStatus: status,
+		responseHeader: headers,
+	}
+	if err := e.miEncode(payload, miRecordSize); err != nil {
+		return nil, err
+	}
+	return e, nil
+}
+
+func (e *Exchange) miEncode(payload []byte, recordSize int) error {
+	var buf bytes.Buffer
+	mi, err := mice.Encode(&buf, payload, recordSize)
+	if err != nil {
+		return err
+	}
+	e.payload = buf.Bytes()
+	e.responseHeader.Add("Content-Encoding", "mi-sha256")
+	e.responseHeader.Add("MI", mi)
+	return nil
+}
+
+// AddSignedHeadersHeader adds 'signed-headers' header to the response.
+//
+// Signed-Headers is a Structured Header as defined by
+// [I-D.ietf-httpbis-header-structure]. Its value MUST be a list (Section 4.8
+// of [I-D.ietf-httpbis-header-structure]) of lowercase strings (Section 4.2 of
+// [I-D.ietf-httpbis-header-structure]) naming HTTP response header fields.
+// Pseudo-header field names (Section 8.1.2.1 of [RFC7540]) MUST NOT appear in
+// this list.
+func (e *Exchange) AddSignedHeadersHeader(ks ...string) {
+	strs := []string{}
+	for _, k := range ks {
+		strs = append(strs, fmt.Sprintf(`"%s"`, strings.ToLower(k)))
+	}
+	s := strings.Join(strs, ", ")
+	e.responseHeader.Add("signed-headers", s)
+}
+
+func (e *Exchange) AddSignatureHeader(s *Signer) error {
+	h, err := s.signatureHeaderValue(e)
+	if err != nil {
+		return err
+	}
+	e.responseHeader.Add("Signature", h)
+	return nil
+}
+
+func (e *Exchange) parseSignedHeadersHeader() []string {
+	unparsed := e.responseHeader.Get("signed-headers")
+
+	rawks := strings.Split(unparsed, ",")
+	ks := make([]string, 0, len(rawks))
+	for _, k := range rawks {
+		ks = append(ks, strings.Trim(k, "\" "))
+	}
+	return ks
+}
+
+func (e *Exchange) encodeCanonicalRequest(enc *cbor.Encoder) error {
+	mes := []*cbor.MapEntryEncoder{
+		cbor.GenerateMapEntry(func(keyE *cbor.Encoder, valueE *cbor.Encoder) {
+			keyE.EncodeByteString([]byte(":method"))
+			valueE.EncodeByteString([]byte("GET"))
+		}),
+		cbor.GenerateMapEntry(func(keyE *cbor.Encoder, valueE *cbor.Encoder) {
+			keyE.EncodeByteString([]byte(":url"))
+			valueE.EncodeByteString([]byte(e.requestUri.String()))
+		}),
+	}
+	return enc.EncodeMap(mes)
+}
+
+func (e *Exchange) encodeResponseHeader(enc *cbor.Encoder, onlySignedHeaders bool) error {
```

Comment:
Be consistent about whether the HTTP header fields are one "header" or a list of "headers". I tend to prefer making it plural, but one can argue that that's not literally correct.

Reply:
Done (I don't have a strong opinion. I adopted plurals).